### ISSUE TITLE
[Question] Setting Prefix for Notes

### ISSUE THREAD

**Issue by jwhonce** [mm21/trilium-alchemy#4]
I am implementing a CLI for TaskManager and I noticed my implementation doesn't set the prefix (TODO or DONE). In JavaScript land 

A snippet of my code:

```python
    task = Note(title="Title", template=template, parents=todo_root, content="Need to...")
    task["attribute"] = "value"

    todo_root += task
    trilium.flush()  # TODO: Is this needed?
    task ^= today
```
I suspect I need to create a `Branch` here but my attempts at that have failed. Notes don't get created if I add `task += Branch()`

I use `api.toggleNoteInParent(true, task.noteId, today.noteId, "TODO")` in JavaScript

Suggestions? TIA

**Comment by mm21** [comment]
Hi,

Correct, you can explicitly create a `Branch` (parent or child) using `task += Branch()` if you want to set a prefix. 

Say you wanted to create a branch adding `task` as a child of `todo_root`, you should be able to use:

```python
task += Branch(parent=todo_root, prefix="TODO")
```

Or:

```python
todo_root += Branch(child=task, prefix="TODO")
```

A more shorthand way of doing that is:

```python
task ^= (todo_root, "TODO") # add parent with prefix
```

Or:

```python
todo_root += (task, "TODO") # add child with prefix
```

In this case you shouldn't pass `parents=todo_root` in `Note()` as that will implicitly create a branch with the same parent/child. A `Note` can be created with no parents initially, it just needs to have at least one by the time it's flushed.

If you did want to keep `parents=todo_root` in `Note()`, another way would be to modify the prefix for the existing branch:

```python
task.branches.parents[0].prefix = "TODO"
```

Or, explicitly lookup the branch first and then set prefix:

```python
branch = task.branches.lookup(todo_root) # lookup parent branch
branch.prefix = "TODO"
```

After this is done, you do need to call `Session.flush()` to commit them. Alternatively `task.flush()` works if you just want to commit that note, not other pending ones.

Can you provide a snippet of where you used `task += Branch()` but the note didn't get created? For more debug info you can print `Session.dirty_set` to see which entities have pending changes. Configuring logging with `logging.basicConfig(level=logging.DEBUG)` also prints more info about which entities are being created when `Session.flush()` is called.

**Comment by jwhonce** [comment]
Thank you!

Two items:

1. For `task ^= (today, "TODO")` pylance reports:
```
Operator "^=" not supported for types "Note" and "tuple[Note, Literal['TODO']]"
  Operator "^" not supported for types "Note" and "tuple[Note, Literal['TODO']]"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
(variable) today: Note
```
2. today += (task, "TODO") fails on an assertion (I had to edit out some data):
```
 /home/jhonce/.local/lib/python3.11/site-packages/trilium_alchemy/core/note/note.py:898 in        │
│ __iadd__                                                                                         │
│                                                                                                  │
│    895 │   │   │   │   # add child note                                                          │
│    896 │   │   │   │   self.branches.children.append(entity)                                     │
│    897 │   │   │   else:                                                                         │
│ ❱  898 │   │   │   │   assert isinstance(entity, Branch)                                         │
│    899 │   │   │   │   branch = entity                                                           │
│    900 │   │   │   │                                                                             │
│    901 │   │   │   │   if branch.parent is None:                                                 │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │ entities = (                                                                                 │ │
│ │            │   Note(title=''│ │
│ │            note_id=2YBIehFtPW35),                                                            │ │
│ │            │   'TODO'                                                                        │ │
│ │            )                                                                                 │ │
│ │   entity = 'TODO'                                                                            │ │
│ │     self = Note(title=11 - Friday, note_id=5vkgtB1YjsmP)                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError
```

**Comment by jwhonce** [comment]
Thanks for the answers!

For `task ^= (today, "TODO")` pylance reports:
```
Operator "^=" not supported for types "Note" and "tuple[Note, Literal['TODO']]"
  Operator "^" not supported for types "Note" and "tuple[Note, Literal['TODO']]"
Pylance[reportGeneralTypeIssues]https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
(variable) today: Note
```
But the branch was created. Yeah!

For `today += (task, "TODO")` an assertion was thrown:
```
 /home/jhonce/.local/lib/python3.11/site-packages/trilium_alchemy/core/note/note.py:898 in        │
│ __iadd__                                                                                         │
│                                                                                                  │
│    895 │   │   │   │   # add child note                                                          │
│    896 │   │   │   │   self.branches.children.append(entity)                                     │
│    897 │   │   │   else:                                                                         │
│ ❱  898 │   │   │   │   assert isinstance(entity, Branch)                                         │
│    899 │   │   │   │   branch = entity                                                           │
│    900 │   │   │   │                                                                             │
│    901 │   │   │   │   if branch.parent is None:                                                 │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │ entities = (                                                                                 │ │
│ │            │   Note(title=''│ │
│ │            note_id=2YBIehFtPW35),                                                            │ │
│ │            │   'TODO'                                                                        │ │
│ │            )                                                                                 │ │
│ │   entity = 'TODO'                                                                            │ │
│ │     self = Note(title=11 - Friday, note_id=5vkgtB1YjsmP)                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError
```

**Comment by mm21** [comment]
Glad `^=` is working! I'll update the type hint to also accept a `tuple[Note, str]`.

I see the problem with `+=`. It's thinking the tuple is an iterable of children, so there needs to be a check to see if it's a tuple first.

There's a test case for `note.children += (child, "prefix")`, so that should be working and do the same thing. I'll fix `note += (child, "prefix")` and add a test case for it, and update the type hint for that too.

Thanks for catching these!

**Comment by mm21** [comment]
Released in `trilium_alchemy-0.1.4`.
